validator/validator_spawner.go

```diff
@@ -0,0 +1,254 @@
+	resolver := func(hash common.Hash) ([]byte, error) {
+		// Check if it's a known preimage
+		if preimage, ok := entry.Preimages[hash]; ok {
+			return preimage, nil
+		}
+		return nil, errors.New("preimage not found")
+	}
+	if err := mach.SetPreimageResolver(resolver); err != nil {
+		return err
+	}
+	err := mach.SetGlobalState(entry.StartState)
+	if err != nil {
+		log.Error("error while setting global state for proving", "err", err, "gsStart", entry.StartState)
+		return fmt.Errorf("error while setting global state for proving: %w", err)
+	}
+	for _, batch := range entry.BatchInfo {
+		err = mach.AddSequencerInboxMessage(batch.Number, batch.Data)
+		if err != nil {
+			log.Error(
+				"error while trying to add sequencer msg for proving",
+				"err", err, "seq", entry.StartState.Batch, "blockNr", entry.Id,
+			)
+			return fmt.Errorf("error while trying to add sequencer msg for proving: %w", err)
+		}
+	}
+	if entry.HasDelayedMsg {
+		err = mach.AddDelayedInboxMessage(entry.DelayedMsgNr, entry.DelayedMsg)
+		if err != nil {
+			log.Error(
+				"error while trying to add delayed msg for proving",
+				"err", err, "seq", entry.DelayedMsgNr, "blockNr", entry.Id,
+			)
+			return fmt.Errorf("error while trying to add delayed msg for proving: %w", err)
+		}
+	}
+	return nil
+}
+
+func (v *ValidationSpawner) ExecuteArbitrator(
+	ctx context.Context, entry *ValidationInput, moduleRoot common.Hash,
+) (GoGlobalState, error) {
+	basemachine, err := v.machineLoader.GetMachine(ctx, moduleRoot, true)
+	if err != nil {
+		return GoGlobalState{}, fmt.Errorf("unabled to get WASM machine: %w", err)
+	}
+
+	mach := basemachine.Clone()
+	err = v.loadEntryToMachine(ctx, entry, mach)
+	if err != nil {
+		return GoGlobalState{}, err
+	}
+	var steps uint64
+	for mach.IsRunning() {
+		var count uint64 = 500000000
+		err = mach.Step(ctx, count)
+		if steps > 0 {
+			log.Debug("validation", "moduleRoot", moduleRoot, "block", entry.Id, "steps", steps)
+		}
+		if err != nil {
+			return GoGlobalState{}, fmt.Errorf("machine execution failed with error: %w", err)
+		}
+		steps += count
+	}
+	if mach.IsErrored() {
+		log.Error("machine entered errored state during attempted validation", "block", entry.Id)
+		return GoGlobalState{}, errors.New("machine entered errored state during attempted validation")
+	}
+	return mach.GetGlobalState(), nil
+}
+
+func (v *ValidationSpawner) ExecuteJit(
+	ctx context.Context, entry *ValidationInput, moduleRoot common.Hash,
+) (GoGlobalState, error) {
+	empty := GoGlobalState{}
+
+	machine, err := v.machineLoader.GetJitMachine(ctx, moduleRoot, true)
+	if err != nil {
+		return empty, fmt.Errorf("unabled to get WASM machine: %w", err)
+	}
+
+	resolver := func(hash common.Hash) ([]byte, error) {
+		// Check if it's a known preimage
+		if preimage, ok := entry.Preimages[hash]; ok {
+			return preimage, nil
+		}
+		return nil, errors.New("preimage not found")
+	}
+	state, err := machine.prove(ctx, entry, resolver)
```

Comment:
This is probably a separate refactor (or will probably be in a future validation service PR), but we shouldn't need the resolver here since we already preloaded all of the preimages into the jit machine. We should also be able to switch arbitrator back to preloading preimages instead of using the callback system which is hacky due to cgo.

Reply:
I think ( need to check) that the preimage resolver is actually not use right now because preimages are loaded. For now, at least, I' like to keep it so if we ever need to bring it back we'll have this code. I think that the complicated parts in in the validators and we're keeping those.